crypto/src/action/output.rs

```diff
@@ -1,26 +1,80 @@
 use ark_ff::UniformRand;
+use bytes::Bytes;
 use rand_core::{CryptoRng, RngCore};
+use std::convert::{TryFrom, TryInto};
 
-use crate::{addresses::PaymentAddress, ka, memo::MemoPlaintext, note, value, Fq, Fr, Note, Value};
+use penumbra_proto::{transaction, Protobuf};
 
-pub const OVK_WRAPPED_LEN_BYTES: usize = 80;
+use crate::{
+    action::{
+        constants::{NOTE_ENCRYPTION_BYTES, OVK_WRAPPED_LEN_BYTES},
+        error::Error,
+    },
+    addresses::PaymentAddress,
+    ka,
+    memo::MemoCiphertext,
+    note,
+    proofs::transparent::{OutputProof, OUTPUT_PROOF_LEN_BYTES},
+    value, Fq, Fr, Note, Value,
+};
 
 pub struct Output {
     pub body: Body,
-    //pub encrypted_memo: MemoCiphertext,
-    pub memo: MemoPlaintext,
-    // Below field is unusued until we implement note encryption.
-    // pub ovk_wrapped_key: [u8; OVK_WRAPPED_LEN_BYTES],
+    pub encrypted_memo: MemoCiphertext,
+    pub ovk_wrapped_key: [u8; OVK_WRAPPED_LEN_BYTES],
+}
+
+impl Protobuf<transaction::Output> for Output {}
+
+impl From<Output> for transaction::Output {
+    fn from(msg: Output) -> Self {
+        transaction::Output {
+            body: Some(msg.body.into()),
+            encrypted_memo: Bytes::copy_from_slice(&msg.encrypted_memo.0),
+            ovk_wrapped_key: Bytes::copy_from_slice(&msg.ovk_wrapped_key),
+        }
+    }
+}
+
+impl TryFrom<transaction::Output> for Output {
+    type Error = Error;
+
+    fn try_from(proto: transaction::Output) -> Result<Self, Self::Error> {
+        if proto.body.is_none() {
+            return Err(Error::ProtobufMissingFieldError);
+        }
+
+        let body = Body::try_from(proto.body.unwrap());
```

Comment:
Rather than unwrapping here, it could be merged with the preceding

```
let body = proto.body.ok_or(Error::OutputBodyMalformed)?
    .try_into()
    .map_err(Error::OutputBodyMalformed)?;
```